### PR TITLE
split keeper registry into proxy and logic

### PR DESCRIPTION
**Pull request by infiloop2**
- Create another contract KeeperRegistryLogic and store it as immutable address on KeeperRegistry
- Moved all storage, errors and events to KeeperRegistryBase so it can be shared within proxy and logic
- Initially moved checkUpkeep to Logic. Along with it I had to move some private functions to the base contract as they were used in both
- Will move rest of the view functions in another PR

### CODE REVIEW

File: contracts/test/v0.8/dev/KeeperRegistry.test.ts [hunk 1]
```diff
@@ -650,11 +676,7 @@ describe('KeeperRegistry', () => {
 
         it('has a large enough gas overhead to cover upkeeps that use all their gas [ @skip-coverage ]', async () => {
           await mock.setCheckGasToBurn(checkGasLimit)
-          await mock.setPerformGasToBurn(executeGas)
-          const gas = checkGasLimit
-            .add(executeGas)
-            .add(PERFORM_GAS_OVERHEAD)
```
Comment (infiloop2) [inline contracts/test/v0.8/dev/KeeperRegistry.test.ts:656]: This is a relic from prev version where we did perform within check

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  /**
+   * @param link address of the LINK Token
+   * @param linkEthFeed address of the LINK/ETH price feed
+   * @param fastGasFeed address of the Fast Gas price feed
+   */
+  constructor(
+    address link,
+    address linkEthFeed,
+    address fastGasFeed
+  ) KeeperRegistryBase(link, linkEthFeed, fastGasFeed) {}
```
Comment (RyanRHall) [inline contracts/src/v0.8/dev/KeeperRegistryLogic.sol:19]: Since these values aren't ever actually used, should we set them as zero addresses?

**Review by RyanRHall — CHANGES_REQUESTED** [review]
Can we move over all of the getter functions also? Everything that is view/external should be good to move.

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol (hunk 2 above)
Comment (RyanRHall) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:19]: Do we need a constructor function on the keeper registry logic contract at all?

File: contracts/src/v0.8/dev/KeeperRegistryBase.sol [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+import "../interfaces/AggregatorV3Interface.sol";
+import "../interfaces/LinkTokenInterface.sol";
+import "../interfaces/KeeperRegistryInterface.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+
+/**
+ * @notice Base Keeper Registry contract, contains shared logic between
+ * KeeperRegistry and KeeperRegistryLogic
+ */
+abstract contract KeeperRegistryBase is ConfirmedOwner, KeeperBase, ReentrancyGuard, Pausable {
```
[the remark this answers is not in the document]
Comment (infiloop2) [reply contracts/src/v0.8/dev/KeeperRegistryBase.sol:17]: yep good idea

File: contracts/src/v0.8/dev/ExecutionPrevention.sol [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+abstract contract ExecutionPrevention {
```
Comment (infiloop2) [inline contracts/src/v0.8/dev/ExecutionPrevention.sol:4]: This is a copy of KeeperBase, renamed to ExecutionPrevention

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol (hunk 2 above)
Comment (infiloop2) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:19]: Yes we need these inside constructor as they are used to set immutable variables which get copied into the code and not storage

Comment (RyanRHall) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:19]: I think we could just do...
```solidity
  constructor() KeeperRegistryBase(address(0), address(0), address(0)) {}
```

wdyt?

**Review by RyanRHall — CHANGES_REQUESTED** [review]

**Review by RyanRHall — APPROVED** [review]

**Pull request merged**